packages/assets-controllers/src/TokensController.test.ts

```diff
@@ -1650,355 +1866,672 @@ describe('TokensController', () => {
+    messenger: approvalControllerMessenger,
+    showApprovalRequest: jest.fn(),
+    typesExcludedFromRateLimiting: [ApprovalType.WatchAsset],
   });
 
-  describe('onTokenListStateChange', () => {
-    it('onTokenListChange', async () => {
-      const stub = stubCreateEthers(tokensController, () => false);
-      await tokensController.addToken({
-        address: '0x01',
-        symbol: 'bar',
-        decimals: 2,
-      });
-      expect(tokensController.state.tokens[0]).toStrictEqual({
-        address: '0x01',
-        decimals: 2,
-        image:
-          'https://static.metafi.codefi.network/api/v1/tokenIcons/1/0x01.png',
-        symbol: 'bar',
-        isERC721: false,
-        aggregators: [],
-        name: undefined,
-      });
+  const controllerMessenger = messenger.getRestricted({
+    name: 'TokensController',
+    allowedActions: [
+      'ApprovalController:addRequest',
+      'NetworkController:getNetworkClientById',
+    ],
+    allowedEvents: [
+      'NetworkController:networkDidChange',
+      'PreferencesController:stateChange',
+      'TokenListController:stateChange',
+    ],
+  });
+  const controller = new TokensController({
+    chainId: ChainId.mainnet,
+    config: {
+      selectedAddress: '0x1',
+      // The tests assume that this is set, but they shouldn't make that
+      // assumption. But we have to do this due to a bug in TokensController
+      // where the provider can possibly be `undefined` if `networkClientId` is
+      // not specified.
+      provider: new FakeProvider(),
+    },
+    messenger: controllerMessenger,
+    ...options,
+  });
 
-      const sampleMainnetTokenList = {
-        '0x01': {
-          address: '0x01',
-          symbol: 'bar',
-          decimals: 2,
-          occurrences: 1,
-          name: 'BarName',
-          iconUrl:
-            'https://static.metafi.codefi.network/api/v1/tokenIcons/1/0x01.png',
-          aggregators: ['Aave'],
-        },
-      };
-      messenger.publish(
-        'TokenListController:stateChange',
-        {
-          tokenList: sampleMainnetTokenList,
-        } as unknown as TokenListState,
-        [],
-      );
+  const triggerPreferencesStateChange = (state: PreferencesState) => {
+    messenger.publish('PreferencesController:stateChange', state, []);
+  };
 
-      expect(tokensController.state.tokens[0]).toStrictEqual({
-        address: '0x01',
-        decimals: 2,
-        image:
-          'https://static.metafi.codefi.network/api/v1/tokenIcons/1/0x01.png',
-        symbol: 'bar',
-        isERC721: false,
-        aggregators: [],
-        name: 'BarName',
-      });
-      stub.restore();
+  const changeNetwork = (providerConfig: ProviderConfig) => {
+    messenger.publish('NetworkController:networkDidChange', {
+      ...defaultNetworkState,
+      providerConfig,
     });
+  };
+
+  const getNetworkClientById = buildMockGetNetworkClientById({
+    // Since we access these networks so often in these tests, register
```

Comment:
We could put these defaults directly in `buildMockGetNetworkClientById`. We do end up needing these in other test files. Thoughts? Would that be too much magic?

Reply:
Hmm. Seems fine to build them directly into `buildMockGetNetworkClientById` to me - I don't think it's too much magic.

Reply:
Okay — done in [9bc9d12](https://github.com/MetaMask/core/pull/4247/commits/9bc9d125ce2a54095b46891fbec6e752cf95ded4).